Source/qol/floatingnumbers.cpp

```diff
@@ -0,0 +1,176 @@
+		num.style |= UiFlags::ColorWhitegold | GetFontSizeByDamage(num.value);
+		break;
+	case FloatingType::DamageAcid:
+		num.style |= UiFlags::ColorDialogYellow | GetFontSizeByDamage(num.value);
+		break;
+	}
+}
+
+} // namespace
+
+void AddFloatingNumber(bool isMyPlayer, Point pos, FloatingType type, int value, int index, UiFlags style)
+{
+	if (!*sgOptions.FloatingNumbers.enableFloatingNumbers)
+		return;
+	if (!*sgOptions.FloatingNumbers.floatingNumbersFromOthers && !isMyPlayer)
+		return;
+	double distanceX = gnScreenWidth * *sgOptions.FloatingNumbers.maxHorizontalDistance / 100.0;
+	double distanceY = gnScreenHeight * *sgOptions.FloatingNumbers.maxVerticalDistance / 100.0;
+	double angle = (500 + rand() % 500) * 0.0062831853;
+	int xCoord = cos(angle) * distanceX;
+	int yCoord = sin(angle) * distanceY;
+	Displacement endOffset = { xCoord, yCoord };
+
+	ClearExpiredNumbers();
+	for (auto &num : FloatingQueue) {
+		if (index != -1 && num.myDmg == isMyPlayer && num.type == type && num.index == index && (SDL_GetTicks() - (int)num.lastMerge <= *sgOptions.FloatingNumbers.mergeTime)) {
+			num.value += value;
+			num.lastMerge = SDL_GetTicks();
+			UpdateFloatingData(num);
+			return;
+		}
+	}
+	FloatingNumber num = FloatingNumber {
+		isMyPlayer, pos, endOffset, "", SDL_GetTicks() + *sgOptions.FloatingNumbers.floatingNumbersLifetime, SDL_GetTicks(), style, type, value, index
+	};
+	UpdateFloatingData(num);
+	num.style |= style; // for handling styles from debug command
+	FloatingQueue.push_back(num);
+}
+
+void DrawFloatingNumbers(const Surface &out)
+{
+	if (!*sgOptions.FloatingNumbers.enableFloatingNumbers)
+		return;
+	ClearExpiredNumbers();
+	for (auto &floatingNum : FloatingQueue) {
+		Point pos = floatingNum.startPos;
+		pos = FloatingCoordsMap[pos.x + pos.y * MAXDUNX];
+		if (pos == Point { 0, 0 })
+			continue;
+		constexpr int tileAndHalf = TILE_HEIGHT * 1.5;
+		pos += { TILE_WIDTH / 2, -tileAndHalf };
+
+		if (!zoomflag) {
+			pos.x *= 2;
+			pos.y *= 2;
+		}
+		pos.x -= GetLineWidth(floatingNum.text) / 2;
+		uint32_t timeLeft = floatingNum.time - SDL_GetTicks();
+		float mul = 1 - (timeLeft / (float)*sgOptions.FloatingNumbers.floatingNumbersLifetime);
+		Point endPos = pos + floatingNum.endOffset * mul;
+		if (endPos.x < 0 || endPos.x >= gnScreenWidth || endPos.y < 0 || endPos.y >= gnScreenHeight)
+			continue;
+		constexpr UiFlags allSizes = UiFlags::FontSize12 | UiFlags::FontSize24 | UiFlags::FontSize30 | UiFlags::FontSize42 | UiFlags::FontSize46;
+		UiFlags shadowStyle = (floatingNum.style & allSizes) | UiFlags::ColorBlack;
+		DrawString(out, floatingNum.text, endPos + Displacement { -1, -1 }, shadowStyle);
+		DrawString(out, floatingNum.text, endPos + Displacement { 1, 1 }, shadowStyle);
+		DrawString(out, floatingNum.text, endPos + Displacement { -1, 1 }, shadowStyle);
+		DrawString(out, floatingNum.text, endPos + Displacement { 1, -1 }, shadowStyle);
+		DrawString(out, floatingNum.text, endPos, floatingNum.style);

```

Comment:
We use CEL for fonts now and CEL has a handy `CelBlitOutlineTo` method that is much faster than rendering 4 times. Let's add outline support to `DrawString` using `CelBlitOutlineTo`?